compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -6604,8 +6604,35 @@ public void visit(BLangOrderByClause orderByClause, AnalyzerData data) {
     }
 
     @Override
-    public void visit(BLangGroupByClause node, AnalyzerData data) {
+    public void visit(BLangGroupByClause groupByClause, AnalyzerData data) {
+        groupByClause.env = data.commonAnalyzerData.queryEnvs.peek();
+        Set<Name> groupingKeys = new HashSet<>();
+        for (BLangGroupingKey groupingKey : groupByClause.groupingKeyList) {
+            if (groupingKey.variableRef != null) {
+                checkExpr(groupingKey.variableRef, groupByClause.env, data);
+                groupingKeys.add(new Name(groupingKey.variableRef.variableName.value));
+            } else {
+                semanticAnalyzer.analyzeNode(groupingKey.variableDef, groupByClause.env,
+                        data.commonAnalyzerData);
+                groupingKeys.add(new Name(groupingKey.variableDef.var.name.value));
+            }
+        }
+        groupByClause.nonGroupingKeys = new HashSet<>();
+        findNonGroupingKeys(groupByClause.env, groupingKeys, groupByClause.nonGroupingKeys);
```

Comment:
We can stop traversing the scopes if we get a semanticError from above typechecking right (line 6612 and 6615)?